FileUpload Android java sample fails with 'java.lang.NoClassDefFoundError' exception

- **OS and version used:** <VERSION>  Ubuntu 18.04 

- **Java runtime used:** <VERSION>  JDK 1.8

- **SDK version used:** <VERSION> iot-device-client:1.19.0

- **Android SDK Version:** <VERSION> 29

  # Description of the issue:
Trying to perform a FileUpload using the android java sample application (available in https://github.com/Azure/azure-iot-sdk-java/tree/master/device/iot-device-samples/android-sample), a systematic exception is hit: 
**`Method threw 'java.lang.NoClassDefFoundError' exception. Cannot evaluate com.microsoft.azure.storage.StorageUri.toString(`**)

# Code sample exhibiting the issue:
  ```
  public CloudBlockBlob(final StorageUri blobAbsoluteUri) throws StorageException {
             super(BlobType.BLOCK_BLOB);

            Utility.assertNotNull("blobAbsoluteUri", blobAbsoluteUri); --> blobAbsoluteUri: Method threw 'java.lang.NoClassDefFoundError' exception. Cannot evaluate com.microsoft.azure.storage.StorageUri.toString()
[...]

```

# Console log of the issue:
`Method threw 'java.lang.NoClassDefFoundError' exception. Cannot evaluate com.microsoft.azure.storage.StorageUri.toString()`

# Solution
To fix the azure-storage package mismatch, app/build.gradle should use these dependencies
```
    api ('com.microsoft.azure.sdk.iot:iot-device-client:1.19.0') {
                exclude module: 'slf4j-api'
                exclude module:'azure-storage'
    }
    implementation 'org.slf4j:slf4j-android:1.7.29'
    implementation 'com.microsoft.azure.android:azure-storage-android:2.0.0'
    implementation 'org.apache.commons:commons-lang3:3.6'
```

instead of simply `api ('com.microsoft.azure.sdk.iot:iot-device-client:1.19.0').`

Note: SLF4 reference enables to get logging information from azure SDK.

****
The ideal solution would be updating build.gradle in android sample application so that users don't waste time investigating this tricky issue;
Exception is not caught in FileUploadTask.java and we issue in "finally" section. Thus, Callback is never received and the file upload request hangs until the timeout...

**Could you please update documentation and fix android sample application?**

Hi Tim, I am seeing a similar Issue in file upload. Again I have used the sample app.. This is the exception log I am seeing. Hope this helps


01-25 16:55:23.809 20167 20167 F linker  : CANNOT LINK EXECUTABLE "/system/bin/dpmd": "/system/lib64/libdpmframework.so" is 32-bit instead of 64-bit
01-25 16:55:24.037 19976 20048 I samplefromazur: Rejecting re-init on previously-failed class java.lang.Class<com.microsoft.azure.storage.core.Utility>: java.lang.NoClassDefFoundError: Failed resolution of: Ljavax/xml/stream/XMLOutputFactory;
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void com.microsoft.azure.storage.core.Utility.<clinit>() (Utility.java:120)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void com.microsoft.azure.storage.core.Utility.assertNotNull(java.lang.String, java.lang.Object) (Utility.java:294)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void com.microsoft.azure.storage.blob.CloudBlockBlob.<init>(com.microsoft.azure.storage.StorageUri) (CloudBlockBlob.java:72)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void com.microsoft.azure.storage.blob.CloudBlockBlob.<init>(java.net.URI) (CloudBlockBlob.java:57)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void com.microsoft.azure.sdk.iot.device.fileupload.FileUploadTask.run() (FileUploadTask.java:156)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at java.lang.Object java.util.concurrent.Executors$RunnableAdapter.call() (Executors.java:458)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void java.util.concurrent.FutureTask.run() (FutureTask.java:266)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run() (ScheduledThreadPoolExecutor.java:301)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void java.util.concurrent.ThreadPoolExecutor.runWorker(java.util.concurrent.ThreadPoolExecutor$Worker) (ThreadPoolExecutor.java:1167)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void java.util.concurrent.ThreadPoolExecutor$Worker.run() (ThreadPoolExecutor.java:641)
01-25 16:55:24.037 19976 20048 I samplefromazur:   at void java.lang.Thread.run() (Thread.java:764)
01-25 16:55:24.037 19976 20048 I samplefromazur: Caused by: java.lang.ClassNotFoundException: Didn't find class "javax.xml.stream.XMLOutputFactory" on path: DexPathList[[zip file "/data/app/com.axesh.iotsamplefromazure-uGyyBVEWV-5sTqdlrBu8OQ==/base.apk"],nativeLibraryDirectories=[/data/app/com.axesh.iotsamplefromazure-uGyyBVEWV-5sTqdlrBu8OQ==/lib/arm64, /system/lib64]]

--- a/device/iot-device-samples/android-sample/app/build.gradle
+++ b/device/iot-device-samples/android-sample/app/build.gradle
@@ -5,7 +5,7 @@ android {
 
     defaultConfig {
         applicationId "com.iothub.azure.microsoft.com.androidsample"
-        minSdkVersion 19
+        minSdkVersion 26
         targetSdkVersion 27
         multiDexEnabled true
         versionCode 1
@@ -44,7 +44,13 @@ dependencies {
     testImplementation 'junit:junit:4.12'
 
     // Remote binary dependency
-    api 'com.microsoft.azure.sdk.iot:iot-device-client:1.20.2'
+    api ('com.microsoft.azure.sdk.iot:iot-device-client:1.20.2') {
+        exclude module: 'slf4j-api'
+        exclude module:'azure-storage'
+    }
+    implementation 'org.slf4j:slf4j-android:1.7.29'
+    implementation 'com.microsoft.azure.android:azure-storage-android:2.0.0'
+    implementation 'org.apache.commons:commons-lang3:3.6'
 }
 
 repositories {

